### PR TITLE
Updating docs for unique key changes in 1.1

### LINKED ISSUE #1105: Support for multiple `unique_key` columns
### Contributions

- [ ] I have read the contribution docs, and understand what's expected of me.

### Link to the page on docs.getdbt.com requiring updates

> updating [these docs](https://docs.getdbt.com/docs/building-a-dbt-project/building-models/configuring-incremental-models#defining-a-uniqueness-constraint-optional) to mention support for multiple unique_key columns (and being aware that support will depend on[ database + strategy](https://docs.getdbt.com/docs/building-a-dbt-project/building-models/configuring-incremental-models#what-is-an-incremental_strategy))

### What part(s) of the page would you like to see updated?

* https://github.com/dbt-labs/dbt-core/pull/4154
* https://github.com/dbt-labs/dbt-core/issues/4738

### Additional information

https://dbt-labs.slack.com/archives/C0131TY7EEA/p1644324413792749

### PR DESCRIPTION
**Pull request by runleonarun**
Closes https://github.com/dbt-labs/docs.getdbt.com/issues/1105
## Pre-release docs
Is this change related to an unreleased version of dbt?
- [ ] Yes: please
    - update the base branch to `next`
    - add Changelog components: `<Changelog>[New/Changed] in v0.x.0</Changelog>`
    - add links to the "New and changed documentation" section of the latest [Migration Guide](../website/docs/docs/guides/migration-guide)
- [ ] No: please ensure the base branch is `current`
- [ ] Unsure: we'll let you know!

## Checklist
If you added new pages (delete if not applicable):
- [ ] The page has been added to `website/sidebars.js`
- [ ] The new page has a unique filename

If you removed existing pages (delete if not applicable):
- [ ] The page has been removed from `website/sidebars.js`
- [ ] An entry has been added to `_redirects`

### CODE REVIEW

File: website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md]: Given the change in v1.1, how about something like:
```suggestion
A `unique_key` is an optional parameter for incremental models, that specifies a field (or combination of fields) that can uniquely identify each row within your model. You can define `unique_key` in a configuration block at the top of your model.
```
We may want to `VersionBlock` that language, and keep the previous language for older versions

File: website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md]:
```suggestion
While some databases support using expressions (eg. `concat(user_id, session_number)`), this syntax is not universally supported. We recommend you pass these columns as a list instead: `unique_key = ['user_id', 'session_number']`.
```
```

File: website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md]: > an update procedure takes place

This differs based on whether the database uses `merge` or `delete+insert`. If the latter, it's not an `update` of the existing row, it's a `delete` of the existing row + insertion of the new row (with the same key)

I think we need to balance the simplicity of the presentation here, while also finding ways to explain nuances like:
- `unique_key` is used by the most common incremental strategies (`delete+insert` + `merge`), but not all. For instance, it's not used by the `insert_overwrite` strategy. We talk about incremental strategies further down; I recently put together [an explainer in our internal Notion](https://www.notion.so/dbtlabs/Incremental-Strategies-573356fb7b2344fe86d3dd3089b8e2b1).
- If `unique_key` is not specified, the behavior is simple "append-only": all rows returned by the model's SQL are inserted into the preexisting table, with no regard for whether they represent dupes. (In `dbt-spark`, this strategy is named `append` explicitly; in other adapters, this is just the behavior of `delete+insert` or `merge` when no `unique_key` is specified.

File: website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md]: I find this sentence a bit confusing. Maybe something more like:
```suggestion
By using `unique_key`, you can ensure that each row from the source table is represented by a single row in your incremental model, without duplicates
```
```

**Review by jtcohen6 — COMMENTED** [review]
@runleonarun Thank you for taking a first stab, and making this happen! Happy to talk through some of the nuances here. The challenge will be, finding a way to present this information that:
- is accessible to newer uses
- is reusable across versions (with VersionBlock additions)
- holds true across all our databases/strategies

**Comment by runleonarun** [comment]
@jtcohen6  Thanks for the reminders/challenges! I will capture the prior to v1.1 functionality in a versioning codeblock, but I might need some help making sure I captured the right content!

File: website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md
Comment (runleonarun) [reply website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md]: Addressed in https://github.com/dbt-labs/docs.getdbt.com/pull/1298/commits/0de7fe73136f06ff9b94ba4152a38897c6e407dd so I could add versioning as well.

File: website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md
Comment (runleonarun) [reply website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md]: @jtcohen6 I think I've addressed these concerns in this commit https://github.com/dbt-labs/docs.getdbt.com/pull/1298/commits/1d3548ebe3c5843b1fb3bface03ba33366d8c54a but please let me know if it needs more tweaking!

File: website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md]: Let's be more prescriptive in our suggested behavior for v1.0 and older:

```suggestion
This optional parameter for incremental models specifies a field that can uniquely identify each row within your model. You can define `unique_key` in a configuration block at the top of your model. If your model doesn't contain a single field that is unique, but rather a combination of columns, we recommend that you create a single column that can serve as unique identifier (by concatenating and hashing those columns), and pass it into your model's configuration.
```

File: website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md]: Starting in v1.1, we can be less prescriptive, and use more neutral language, since both one column and multiple columns in combination work well. The thing to feel strongly about is the `:::tip` you called out below.
```suggestion
The `unique_key` should be supplied in your model definition as a string representing a simple column or a list of single quoted column names that can be used together, for example, `['col1', 'col2', …])`.
```

File: website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md]: I agree with the first sentence; I'd prefer to situate it in the specific circumstance we're discussing
```suggestion
In cases where you need multiple columns in combination to uniquely identify each row, we recommend you pass these columns as a list (`unique_key = ['user_id', 'session_number']`), rather than a string expression (`unique_key = 'concat(user_id, session_number)'`).

By using the first syntax, which is more universal, dbt can ensure that the columns will be templated into your incremental model materialization in a way that's appropriate to your database.
```

File: website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md]: 👍 love this, I think we can give just a tiny bit more color:
```suggestion
While common incremental strategies, such as`delete+insert` + `merge`, might use `unique_key`, others don't. For example, the `insert_overwrite` strategy does not use `unique_key`, because it operates on partitions of data rather than individual rows. For more information, see [About incremental_strategy](#about-incremental_strategy).
```

File: website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md]: This is close. The references "columns" and "rows" feel a bit confusing to me here. I think we should zoom this up to be at a _slightly_ higher conceptual level, now that we've developed the `unique_key` concept above
```
When a `unique_key` is defined, for each row of "new" data returned by your dbt model:
* If the same `unique_key` is present in in the "new" and "old" model data, dbt will update/replace the old row with the new row of data. The exact mechanics of that update/replace takes place will vary depending on your database and [incremental strategy](#about-incremental_strategy).
* If the `unique_key` is _not_ present in the "old" data, dbt will insert the entire row into the table.
```

**Review by jtcohen6 — APPROVED** [review]
Super super close! I think all the pieces are in place. Comments from me are just small recommends to firm up the final language. Approving so as not to block merge when you feel it's ready

File: website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md
Comment (runleonarun) [reply website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md]: Thanks Jeremy! This context helps.

File: website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md
Comment (runleonarun) [reply website/docs/docs/building-a-dbt-project/building-models/configuring-incremental-models.md]: Addressed in this commit: https://github.com/dbt-labs/docs.getdbt.com/pull/1298/commits/838b1da8f9687ae35e6dfb09955c791052431fe1

**Pull request merged**
